File: leetcode/leetcode-34.py
from typing import List

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def lower_bound():
            l = 0
            r = len(nums)-1
            if r < 0:
                return -1
            while l <= r:
                mid = (l+r)//2
                if target > nums[mid]:
                    l = mid+1
                else:
                    r = mid-1
            return l if 0<=l<len(nums) and nums[l] == target else -1
        def upper_bound():
            l = 0
            r = len(nums)-1
            if r < 0:
                return -1
            while l <= r:
                mid = (l+r)//2
                if target >= nums[mid]:
                    l = mid+1
                else:
                    r = mid-1
            return r if 0 <=r< len(nums)  and nums[r] == target else -1
        ans = [lower_bound(), upper_bound()]
        return ans


if __name__ == "__main__":
    so = Solution()
    print(so.searchRange([2, 2], 3))

File: leetcode/leetcode-34.cpp
/*
 *@File Name          :leetcode-34.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/7/19 10:06
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "algorithm"
#include "vector"
using namespace std;

class Solution {
 public:
  int serachL(vector<int>& nums, int target) {
    int l = 0, r = nums.size() - 1;
    if (r < 0) return -1;

    while (l < r) {
      int mid = (l + r) / 2;
      if (nums[mid] >= target)
        r = mid;
      else
        l = mid + 1;
    }
    return nums[l] == target ? l : -1;
  }

  int serachR(vector<int>& nums, int target) {
    int l = 0, r = nums.size() - 1;
    if (r < 0) return -1;
    while (l < r) {
      int mid = (l + r + 1) / 2;
      if (nums[mid] <= target)
        l = mid;
      else
        r = mid - 1;
    }
    return nums[l] == target ? l : -1;
  }

  vector<int> searchRange(vector<int>& nums, int target) {
    return {serachL(nums, target), serachR(nums, target)};
  }
};